File: frontend/src/lib/utils/naismith_674.test.ts
// TDD RED — Issue #674 Fahrradtour-Aktivitätstyp (TypeScript)
//
// Spec: docs/specs/modules/issue_674_aktivitaetstyp_fahrrad.md
// Workflow: phase5_tdd_red — Tests MÜSSEN FEHLSCHLAGEN bis Phase 6.
//
// RED-Ursache: activityToSpeed() existiert nicht in naismith.ts → Import-Fehler.
//
// Ausführung:
//   cd frontend && node --experimental-strip-types --test \
//     src/lib/utils/naismith_674.test.ts

import { test } from 'node:test';
import assert from 'node:assert/strict';

// activityToSpeed existiert NOCH NICHT → Import-Fehler = RED
import { activityToSpeed, computeArrivalTimes } from './naismith.ts';
import type { Stage, Waypoint, ActivityType } from '../types.ts';

// =============================================================================
// Helpers
// =============================================================================

function wp(id: string, lat: number, lon: number, elevation_m: number): Waypoint {
	return { id, name: id, lat, lon, elevation_m };
}

function stage(id: string, waypoints: Waypoint[], start_time?: string): Stage {
	const s: Stage = { id, name: id, date: '2026-06-09', waypoints };
	if (start_time !== undefined) s.start_time = start_time;
	return s;
}

// 0.17987° Lat ≈ 20 km bei ~42° N
const DELTA_20KM_LAT = 0.17987;

// =============================================================================
// AC-7: activityToSpeed — Geschwindigkeit aus ActivityType
// =============================================================================

test('AC-7: activityToSpeed("fahrrad_25") === 25', () => {
	assert.equal(activityToSpeed('fahrrad_25'), 25.0);
});

test('AC-7: activityToSpeed("fahrrad_20") === 20', () => {
	assert.equal(activityToSpeed('fahrrad_20'), 20.0);
});

test('AC-7: activityToSpeed("fahrrad_15") === 15', () => {
	assert.equal(activityToSpeed('fahrrad_15'), 15.0);
});

test('AC-7: activityToSpeed(undefined) === 4 (Wanderer-Default bleibt erhalten)', () => {
	assert.equal(activityToSpeed(undefined), 4.0);
});

test('AC-7: activityToSpeed("trekking") === 4 (Wander-Aktivitäten → 4 km/h)', () => {
	assert.equal(activityToSpeed('trekking' as ActivityType), 4.0);
});

test('AC-7: activityToSpeed("skitour") === 4 (Ski → ebenfalls Wandergeschwindigkeit)', () => {
	assert.equal(activityToSpeed('skitour' as ActivityType), 4.0);
});

// =============================================================================
// AC-1 (TS-Seite): computeArrivalTimes mit Fahrrad-Speed
// =============================================================================

test('AC-1 TS: Fahrrad 20 km/h, 20 km flach → 08:00→09:00', () => {
	const s = stage(
		's674',
		[
			wp('w1', 42.0, 9.0, 200),
			wp('w2', 42.0 + DELTA_20KM_LAT, 9.0, 200),
		],
		'08:00'
	);
	// computeArrivalTimes wird mit Speed 20 km/h aufgerufen
	const arrivals = computeArrivalTimes(s, '08:00', activityToSpeed('fahrrad_20'));
	assert.equal(arrivals.length, 2);
	assert.equal(arrivals[0], '08:00');
	assert.equal(arrivals[1], '09:00', '20 km ÷ 20 km/h = 1 h → +1 h');
});

test('AC-2 TS: Wanderer-Default (kein Speed) → 20 km braucht 5 h → 13:00', () => {
	const s = stage(
		's674-wander',
		[
			wp('w1', 42.0, 9.0, 200),
			wp('w2', 42.0 + DELTA_20KM_LAT, 9.0, 200),
		],
		'08:00'
	);
	const arrivals = computeArrivalTimes(s, '08:00'); // kein Speed → Default 4 km/h
	assert.equal(arrivals.length, 2);
	assert.equal(arrivals[1], '13:00', '20 km ÷ 4 km/h = 5 h → 13:00');
});

test('AC-5 TS: Fahrrad 25 km/h kommt früher an als Wanderer', () => {
	const waypoints = [
		wp('w1', 42.0, 9.0, 200),
		wp('w2', 42.0 + DELTA_20KM_LAT, 9.0, 200),
	];
	const s = stage('s', waypoints, '08:00');

	const wanderer = computeArrivalTimes(s, '08:00');
	const fahrrad = computeArrivalTimes(s, '08:00', activityToSpeed('fahrrad_25'));

	// Fahrrad muss früher ankommen als zu Fuss
	const parse = (t: string) => parseInt(t.slice(0, 2)) * 60 + parseInt(t.slice(3));
	assert.ok(
		parse(fahrrad[1]) < parse(wanderer[1]),
		`Fahrrad (${fahrrad[1]}) soll früher als Wanderer (${wanderer[1]}) ankommen`
	);
});

File: internal/model/naismith_674_test.go
package model

import "testing"

// TDD RED — Issue #674 Fahrradtour-Aktivitätstyp
//
// Spec: docs/specs/modules/issue_674_aktivitaetstyp_fahrrad.md
// Workflow: phase5_tdd_red — Tests MÜSSEN FEHLSCHLAGEN bis Phase 6.
//
// KEINE MOCKS — echte Structs, echte Berechnung.
//
// RED-Ursache: ActivitySpeed(string) gibt es nicht → Compile-Fehler.

// TestActivitySpeed_Fahrrad25 prüft AC-6: ActivitySpeed("fahrrad_25")
// liefert FlatKmh=25, AscentMh=600, DescentMh=1000.
func TestActivitySpeed_Fahrrad25(t *testing.T) {
	s := ActivitySpeed("fahrrad_25") // nicht vorhanden → Compile-Fehler = RED
	if s.FlatKmh != 25.0 {
		t.Fatalf("FlatKmh = %v, want 25.0", s.FlatKmh)
	}
	if s.AscentMh != 600.0 {
		t.Fatalf("AscentMh = %v, want 600.0", s.AscentMh)
	}
	if s.DescentMh != 1000.0 {
		t.Fatalf("DescentMh = %v, want 1000.0", s.DescentMh)
	}
}

// TestActivitySpeed_Fahrrad20 prüft AC-6 für fahrrad_20.
func TestActivitySpeed_Fahrrad20(t *testing.T) {
	s := ActivitySpeed("fahrrad_20")
	if s.FlatKmh != 20.0 {
		t.Fatalf("FlatKmh = %v, want 20.0", s.FlatKmh)
	}
	if s.AscentMh != 600.0 {
		t.Fatalf("AscentMh = %v, want 600.0", s.AscentMh)
	}
}

// TestActivitySpeed_Fahrrad15 prüft AC-6 für fahrrad_15.
func TestActivitySpeed_Fahrrad15(t *testing.T) {
	s := ActivitySpeed("fahrrad_15")
	if s.FlatKmh != 15.0 {
		t.Fatalf("FlatKmh = %v, want 15.0", s.FlatKmh)
	}
	if s.AscentMh != 600.0 {
		t.Fatalf("AscentMh = %v, want 600.0", s.AscentMh)
	}
}

// TestActivitySpeed_Default prüft AC-6: ActivitySpeed("") liefert Wanderer-Default
// (FlatKmh=4, AscentMh=300, DescentMh=500).
func TestActivitySpeed_Default(t *testing.T) {
	s := ActivitySpeed("")
	if s.FlatKmh != 4.0 {
		t.Fatalf("FlatKmh = %v, want 4.0 (Wanderer-Default)", s.FlatKmh)
	}
	if s.AscentMh != 300.0 {
		t.Fatalf("AscentMh = %v, want 300.0", s.AscentMh)
	}
	if s.DescentMh != 500.0 {
		t.Fatalf("DescentMh = %v, want 500.0", s.DescentMh)
	}
}

// TestActivitySpeed_UnknownActivity prüft: unbekannter Aktivitätswert → Wanderer-Default.
func TestActivitySpeed_UnknownActivity(t *testing.T) {
	s := ActivitySpeed("paragliding") // unbekannt → Default
	if s.FlatKmh != 4.0 {
		t.Fatalf("FlatKmh = %v, want 4.0 (Default für unbekannte Aktivität)", s.FlatKmh)
	}
}

// TestComputeStageArrivals_Fahrrad20_Flat prüft AC-1:
// Trip activity="fahrrad_20", 2 Wegpunkte ~20 km flach, start_time="08:00"
// → wp[1].ArrivalCalculated == "09:00" (20 km ÷ 20 km/h = 1 h).
//
// 0.17987° Lat ≈ 20 km bei ~42° N (5 × DELTA_4KM_LAT aus naismith_test.go).
func TestComputeStageArrivals_Fahrrad20_Flat(t *testing.T) {
	stage := &Stage{
		ID: "S674-1", Name: "Fahrrad-Etappe", Date: "2026-06-09",
		StartTime: strPtr("08:00"),
		Waypoints: []Waypoint{
			{ID: "W1", Name: "Start", Lat: 42.0, Lon: 9.0, ElevationM: 200},
			{ID: "W2", Name: "Ziel", Lat: 42.17987, Lon: 9.0, ElevationM: 200}, // ~20 km
		},
	}
	speeds := ActivitySpeed("fahrrad_20")
	ComputeStageArrivals(stage, speeds) // neue Signatur → Compile-Fehler = RED

	if stage.Waypoints[1].ArrivalCalculated == nil {
		t.Fatal("wp[1] ArrivalCalculated == nil, want gesetzt")
	}
	got := *stage.Waypoints[1].ArrivalCalculated
	if got != "09:00" {
		t.Fatalf("wp[1] arrival = %q, want 09:00 (20 km ÷ 20 km/h = 1 h)", got)
	}
}

// TestComputeStageArrivals_Fahrrad15_Ascent prüft AC-3:
// Trip activity="fahrrad_15", +600 m Aufstieg ~0 km Horizontal, start_time="08:00"
// → wp[1].ArrivalCalculated == "09:00" (600 m ÷ 600 Hm/h = 1 h).
func TestComputeStageArrivals_Fahrrad15_Ascent(t *testing.T) {
	stage := &Stage{
		ID: "S674-3", Name: "Berg-Etappe", Date: "2026-06-09",
		StartTime: strPtr("08:00"),
		Waypoints: []Waypoint{
			{ID: "W1", Name: "Tal", Lat: 42.0, Lon: 9.0, ElevationM: 200},
			{ID: "W2", Name: "Gipfel", Lat: 42.0, Lon: 9.0, ElevationM: 800}, // +600 m, ~0 km
		},
	}
	speeds := ActivitySpeed("fahrrad_15")
	ComputeStageArrivals(stage, speeds)

	if stage.Waypoints[1].ArrivalCalculated == nil {
		t.Fatal("wp[1] ArrivalCalculated == nil, want gesetzt")
	}
	got := *stage.Waypoints[1].ArrivalCalculated
	if got != "09:00" {
		t.Fatalf("wp[1] arrival = %q, want 09:00 (600 m ÷ 600 Hm/h = 1 h)", got)
	}
}

// TestComputeStageArrivals_WandererDefault_Unchanged prüft AC-2 + AC-8:
// Leere activity → Wanderer-Default 4 km/h → identische Ergebnisse wie vor Refaktorierung.
// 0.036° lat ≈ 4 km → wp[1] = "09:00" (4 km ÷ 4 km/h = 1 h).
func TestComputeStageArrivals_WandererDefault_Unchanged(t *testing.T) {
	stage := &Stage{
		ID: "S674-2", Name: "Wander-Etappe", Date: "2026-06-09",
		StartTime: strPtr("08:00"),
		Waypoints: []Waypoint{
			{ID: "W1", Name: "Start", Lat: 47.000, Lon: 11.0, ElevationM: 1000},
			{ID: "W2", Name: "Ziel", Lat: 47.036, Lon: 11.0, ElevationM: 1000},
		},
	}
	speeds := ActivitySpeed("") // leer → Wanderer-Default
	ComputeStageArrivals(stage, speeds)

	if stage.Waypoints[1].ArrivalCalculated == nil {
		t.Fatal("wp[1] ArrivalCalculated == nil, want gesetzt")
	}
	got := *stage.Waypoints[1].ArrivalCalculated
	if got != "09:00" {
		t.Fatalf("wp[1] arrival = %q, want 09:00 (Wanderer 4 km/h, 4 km ÷ 4 = 1 h)", got)
	}
}

// TestFahrrad20_FasterThanWanderer prüft AC-5 (Backend-Seite):
// Fahrradtour kommt früher an als Wanderer.
// Beide starten 08:00, flache 20 km. Wanderer: 5 h → 13:00. Fahrrad 20: 1 h → 09:00.
func TestFahrrad20_FasterThanWanderer(t *testing.T) {
	makeStage := func(id string) *Stage {
		return &Stage{
			ID: id, Name: id, Date: "2026-06-09",
			StartTime: strPtr("08:00"),
			Waypoints: []Waypoint{
				{ID: "W1", Name: "Start", Lat: 42.0, Lon: 9.0, ElevationM: 200},
				{ID: "W2", Name: "Ziel", Lat: 42.17987, Lon: 9.0, ElevationM: 200},
			},
		}
	}

	stageWander := makeStage("wander")
	stageFahrrad := makeStage("fahrrad")

	ComputeStageArrivals(stageWander, ActivitySpeed(""))
	ComputeStageArrivals(stageFahrrad, ActivitySpeed("fahrrad_20"))

	wander := *stageWander.Waypoints[1].ArrivalCalculated
	fahrrad := *stageFahrrad.Waypoints[1].ArrivalCalculated

	wMins := hhmmToMinutes(t, wander)
	fMins := hhmmToMinutes(t, fahrrad)

	if fMins >= wMins {
		t.Fatalf("Fahrrad (%q = %d min) soll früher als Wanderer (%q = %d min) ankommen",
			fahrrad, fMins, wander, wMins)
	}
}
